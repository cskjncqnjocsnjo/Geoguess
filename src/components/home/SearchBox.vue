<template>
    <div class="search-box">
        <h2>{{ $t('Home.record') }}: {{ record / 1000 }} km</h2>
        <div class="search-box__search-bar">
            <v-combobox
                :items="items"
                :search-input.sync="search"
                :loading="isLoading"
                autofocus
                :placeholder="
                    isValidGeoJson
                        ? $t('Home.searchBar.customLoaded')
                        : $t('Home.searchBar.enterCity')
                "
                v-model="place"
                :disabled="isValidGeoJson"
                :persistent-hint="isValidGeoJson"
                :background-color="isValidGeoJson ? 'primary' : 'secondary'"
                append-icon="mdi-magnify"
                dark
                rounded
                height="50"
                full-width
                id="search-input"
            >
            </v-combobox>

            <v-btn
                icon
                class="btn-customs"
                color="primary"
                @click="dialogCustom = !dialogCustom"
                height="50"
            >
                <v-icon>mdi-map-plus</v-icon>
            </v-btn>
        </div>
        <DialogCustomMap
            :visibility="dialogCustom"
            @change-visibility="dialogCustom = !dialogCustom"
        />

        <div class="search-box__btns">
            <DialogRoom singlePlayer :place="place" :geoJson="geoJson" />

            <DialogRoom :place="place" :geoJson="geoJson" />
        </div>
    </div>
</template>
<script>
import axios from 'axios';
import DialogCustomMap from '@/components/home/DialogCustomMap';
import DialogRoom from '@/components/widgets/dialog/DialogRoom';
import { mapGetters } from 'vuex';
export default {
    components: {
        DialogRoom,
        DialogCustomMap,
    },
    data() {
        return {
            record: localStorage.getItem('record'),
            place: '',
            dialog: false,
            entries: [],
            isLoading: false,
            search: '',
            dialogCustom: false,
        };
    },

    computed: {
        ...mapGetters(['isValidGeoJson', 'geoJson']),
        items() {
            return this.entries.map((entry) => entry.properties.name);
        },
    },

    watch: {
        geoJson(val) {
            if (val !== null) {
                this.place = '';
            }
        },
        search(val) {
            // Items have already been requested
            if (!val) return;

            this.isLoading = true;
            // Lazily load input items
            axios
                .get(`https://photon.komoot.de/api/?q=${encodeURI(val)}`)
                .then((res) => {
                    if (res.status === 200 && res.data && res.data.features) {
                        this.entries = res.data.features.filter(
                            (node) => node.properties.osm_type === 'R'
                        );
                    }
                })
                .catch(() => {})
                .finally(() => (this.isLoading = false));
        },
    },
};
</script>
<style lang="scss">
.search-box {
    h2 {
        text-align: center;
    }
    .search-box__search-bar {
        display: flex;
        .btn-customs {
            margin-top: 18px;
        }
    }
    .v-input {
        font-size: 1.2rem !important;
    }
    .search-box__btns {
        display: flex;
        justify-content: space-around;
    }
}
@media (max-width: 380px) {
    .search-box {
        .v-input {
            font-size: 1rem !important;
        }

        .search-box__search-bar {
            flex-direction: column;
            width: 100vw;
            margin: auto;
            .btn-customs {
                margin: 2% auto;
            }
        }
        .search-box__btns {
            flex-direction: column;
            .v-btn {
                width: 80%;
                margin: 2% auto;
            }
        }
    }
}
</style>
