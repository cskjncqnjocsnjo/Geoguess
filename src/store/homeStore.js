import * as MutationTypes from './mutation-types';
import { validURL } from '@/utils';

import axios from 'axios';
import { isGeoJSONValid } from '../utils';
export default {
    state: () => ({
        geojson: null,
        listMaps: [],
        openDialogSinglePlayer: false,
        openDialogMultiPlayer: false,
    }),
    mutations: {
        [MutationTypes.HOME_SET_GEOJSON](state, geojson) {
            state.geojson = geojson;
        },
        [MutationTypes.HOME_SET_LISTMAPS](state, listMaps) {
            state.listMaps = listMaps;
        },
        [MutationTypes.HOME_SET_SINGLEPLAYER](state, status) {
            state.openDialogSinglePlayer = status;
        },
        [MutationTypes.HOME_SET_MULTIPLAYER](state, status) {
            state.openDialogMultiPlayer = status;
        },
    },

    getters: {
        geoJsonString(state) {
            if (!state.geojson) {
                return '';
            }
            return JSON.stringify(state.geojson, null, 2);
        },
        geoJson(state) {
            return state.geojson;
        },
        isValidGeoJson(state) {
            if (!state.geojson) {
                return null;
            }
            return isGeoJSONValid(state.geojson);
        },
        maps(state) {
            return state.listMaps;
        },
    },

    actions: {
        async loadGeoJsonFromUrl({ commit }, url) {
            if (validURL(url)) {
                // if gist url get raw
                if (RegExp('^(https?://)?gist.github.com/').test(url)) {
                    let urlSplit = url.split('/');
                    if (
                        urlSplit.length > 3 &&
                        urlSplit[urlSplit.length - 1] !== 'raw'
                    ) {
                        urlSplit[urlSplit.length - 3] =
                            'gist.githubusercontent.com';
                        urlSplit.push('raw');
                        url = urlSplit.join('/');
                    }
                }
                const geojson = await axios
                    .get(url)
                    .then((res) => {
                        if (res.status === 200 && res.data) {
                            if (typeof res.data === 'object') {
                                return res.data;
                            } else {
                                return JSON.parse(res.data);
                            }
                        }
                    })
                    .catch((err) => {
                        // eslint-disable-next-line no-console
                        console.log(err);
                    });

                commit(MutationTypes.HOME_SET_GEOJSON, geojson);
            }
        },
        setGeoJson({ commit }, geojson) {
            commit(MutationTypes.HOME_SET_GEOJSON, geojson);
        },
        setGeoJsonString({ commit }, geojson) {
            let obj = null;
            if (geojson !== '') {
                obj = JSON.parse(geojson);
            }
            commit(MutationTypes.HOME_SET_GEOJSON, obj);
        },
        async getListMaps({ commit }) {
            const maps = await axios
                .get(
                    process.env.VUE_APP_LIST_MAPS_JSON_URL ||
                        'https://raw.githubusercontent.com/GeoGuess/GeoGuess-Maps/main/maps.json'
                )
                .then((res) => res.data.maps);

            commit(MutationTypes.HOME_SET_LISTMAPS, maps);
        },
        playSinglePlayer({ commit }) {
            commit(MutationTypes.HOME_SET_SINGLEPLAYER, true);
        },
        playMultiPlayer({ commit }) {
            commit(MutationTypes.HOME_SET_MULTIPLAYER, true);
        },
        resetSinglePlayer({ commit }) {
            commit(MutationTypes.HOME_SET_SINGLEPLAYER, false);
        },
        resetMultiPlayer({ commit }) {
            commit(MutationTypes.HOME_SET_MULTIPLAYER, false);
        },
    },
};
