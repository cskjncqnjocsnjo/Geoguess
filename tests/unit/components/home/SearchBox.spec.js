jest.mock('axios', () => {
    const responseTls = {
        features: [
            {
                geometry: {
                    coordinates: [1.4442469, 43.6044622],
                    type: 'Point',
                },
                type: 'Feature',
                properties: {
                    osm_id: 35738,
                    osm_type: 'R',
                    extent: [1.3503956, 43.668708, 1.5153795, 43.532654],
                    country: 'France',
                    osm_key: 'place',
                    osm_value: 'city',
                    name: 'Toulouse',
                    state: 'Occitania',
                },
            },
            {
                geometry: {
                    coordinates: [1.4455249, 43.5271458],
                    type: 'Point',
                },
                type: 'Feature',
                properties: {
                    osm_id: 158980,
                    osm_type: 'R',
                    extent: [1.4202811, 43.546134, 1.4534675, 43.5126463],
                    country: 'France',
                    osm_key: 'place',
                    osm_value: 'village',
                    postcode: '31320',
                    name: 'Vieille-Toulouse',
                    state: 'Occitania',
                },
            },
            {
                geometry: {
                    coordinates: [5.587384, 46.8232142],
                    type: 'Point',
                },
                type: 'Feature',
                properties: {
                    osm_id: 156434,
                    osm_type: 'Z',
                    extent: [5.5693913, 46.8364042, 5.5996211, 46.8079512],
                    country: 'France',
                    osm_key: 'place',
                    osm_value: 'village',
                    postcode: '39230',
                    name: 'Toulouse-le-Château',
                    state: 'Bourgogne-Franche-Comté',
                },
            },
        ],
    };

    return {
        get: jest.fn(() =>
            Promise.resolve({
                status: 200,
                data: responseTls,
            })
        ),
    };
});

import SearchBox from '@/components/home/SearchBox.vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import appInit from '../../utils/appInit';
import axios from 'axios';

import Vuex from 'vuex';
import homeStore from '../../../../src/store/homeStore';

const args = appInit(createLocalVue());
describe('SearchBox.vue', () => {
    let store;
    beforeEach(() => {
        store = new Vuex.Store({
            modules: {
                homeStore: {
                    state: homeStore.state,
                    getters: homeStore.getters,
                },
            },
        });
    });
    it('test search Input', () => {
        const wrapper = shallowMount(SearchBox, { ...args, store });
        wrapper.setData({ search: 'Toulouse' });

        expect(wrapper.exists('#search-input'));
        wrapper.vm.$nextTick(() => {
            expect(axios.get).toBeCalledWith(
                'https://photon.komoot.de/api/?q=Toulouse'
            );
            wrapper.vm.$nextTick(() => {
                expect(wrapper.vm.entries).toHaveLength(2);
                expect(wrapper.vm.items).toHaveLength(2);
                expect(wrapper.vm.items[0]).toEqual('Toulouse');
            });
        });

        expect(wrapper).toMatchSnapshot();
    });
});
