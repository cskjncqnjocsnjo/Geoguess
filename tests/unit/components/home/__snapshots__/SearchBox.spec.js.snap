// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox.vue test search Input 1`] = `
<div class="search-box">
  <h2>Record: 0 km</h2>
  <div class="search-box__search-bar">
    <v-combobox-stub dark="true" errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" appendicon="mdi-magnify" backgroundcolor="secondary" height="50" id="search-input" loaderheight="2" autofocus="true" clearicon="$clear" fullwidth="true" placeholder="Enter city, state or country" rounded="true" type="text" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" nodatatext="$vuetify.noDataText" items="" itemcolor="primary" itemdisabled="disabled" itemtext="text" itemvalue="value" menuprops="[object Object]" returnobject="true" allowoverflow="true" filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }" searchinput="" delimiters=""></v-combobox-stub>
    <v-btn-stub color="primary" height="50" tag="button" activeclass="" icon="true" type="button" class="btn-customs">
      <v-icon-stub>mdi-map-plus</v-icon-stub>
    </v-btn-stub>
  </div>
  <dialogcustommap-stub></dialogcustommap-stub>
  <div class="search-box__btns">
    <dialogroom-stub place="" singleplayer="true"></dialogroom-stub>
    <dialogroom-stub place=""></dialogroom-stub>
  </div>
</div>
`;
